src/manifolds/ProbabilitySimplex.jl

```diff
@@ -0,0 +1,294 @@
+`p`. The formula reads
+````math
+g_p(X,Y) = \sum_{i=1}^{n+1}\frac{X_iY_i}{p}
+````
+"""
+function inner(::ProbabilitySimplex, p, X, Y)
+    d = zero(Base.promote_eltype(p,X,Y))
+    @inbounds for i in eachindex(p,X,Y)
+        d += X[i]*Y[i]/p[i]
+    end
+    return d
+end
+
+@doc raw"""
+    inverse_retract(M::ProbabilitySimplex, p, q, ::SoftmaxInverseRetraction)
+
+Compute a first order approximation by projection. The formula reads
+````math
+\operatorname{retr}^{-1}_p q = \bigl( I - \frac{1}{n}\mathbb{1}\mathbb{1}^\mathrm{T}\bigr)(\log(q)-\log(p))
+````
+where $\mathbb{1}$ is the column vector containing ones and $\log$ is applied elementwise.
+"""
+inverse_retract(::ProbabilitySimplex, ::Any, ::Any, ::SoftmaxInverseRetraction)
+
+function inverse_retract!(::ProbabilitySimplex{n}, X, p, q, ::SoftmaxInverseRetraction) where {n}
+    X .= (one(zeros(n+1,n+1)) - 1/n.*ones(n+1,n+1) ) * (log.(q) - log.(p))
+    return X
+end
+
+
+@doc raw"""
+    log(M::ProbabilitySimplex, p, q)
+
+Compute the logarithmic map of `p` and `q` on the [`ProbabilitySimplex`](@ref) `M`.
+
+````math
+\log_pq = \frac{d_{Δ^n}(p,q)}{\sqrt{1-⟨\sqrt{p},\sqrt{q}⟩}}(\sqrt{pq} - ⟨\sqrt{p},\sqrt{q}⟩p),
+````
+
+where $pq$ and $\sqrt{p}$ is meant elementwise.
+"""
+log(::ProbabilitySimplex, ::Any...)
+
+function log!(M::ProbabilitySimplex, X, p, q)
+    s = dot(sqrt.(p),sqrt.(q))
+    X .= (distance(M,p,q)/(1-s^2)) .* ( sqrt.(p.*q) .- s.*p )
+    return X
+end
+
+@doc raw"""
+    manifold_dimension(M::ProbabilitySimplex{n})
+
+Returns the manifodl dimension of the probability siomplex in $ℝ^{n+1}$, i.e.
+````math
+    \dim_{Δ^n} = n.
+````
+"""
+manifold_dimension(::ProbabilitySimplex{n}) where {n} = n
+
+@doc raw"""
+    project(M::ProbabilitySimplex, p, Y)
+
+project `Y` from the embedding onto the tangent space at `p` on
+the [`ProbabilitySimplex`](@ref) `M`. The formula reads
+
+````math
+\operatorname{proj}_{Δ^n}(p,Y) = p\bigl(Y - ⟨p,Y⟩\mathbb{1}),
+````
+
+where multiplication is meant elementwise and $\mathbb{1}$ is the vector of ones.
+"""
+project(::ProbabilitySimplex, ::Any, ::Any)
+
+project!(::ProbabilitySimplex, X, p, Y) = X .= (p.*( Y .- dot(p,Y)))
+
+@doc raw"""
+    representation_size(::ProbabilitySimplex{n})
+
+return the representation size of points in the $n$-dimensional probability simplex,
+i.e. an array size of `(n+1,)`.
+"""
+representation_size(::ProbabilitySimplex{n}) where {n} = (n+1,)
+
+@doc raw"""
+    retract(M::ProbabilitySimplex, p, X, ::SoftmaxRetraction)
+
+Compute a first order approximation by applying the softmax function. The formula reads
+
+````math
+\operatorname{retr}_p X = \frac{p\mathrm{e}^X}{⟨p,\mathrm{e}^X⟩},
+````
+
+where multiplication, exponentiation and division are meant elementwise.
+"""
+retract(::ProbabilitySimplex, ::Any, ::Any, ::SoftmaxRetraction)
+
+function retract!(::ProbabilitySimplex, q, p, X, ::SoftmaxRetraction)
+    q .= p.*exp.(X)./(dot(p,exp.(X)))
```

Comment:
Haven't tested this but should be more efficient

```suggestion
    s = zero(eltype(q))
    @inbounds for i in eachindex(q, p, X)
        q[i] = p[i] * exp(X[i])
        s += q[i]
    end
    q ./= s
```

Reply:
I've checked your version for 3-element vectors and it's about two times faster.